strawberryfields/circuitspecs/chip2.py

```diff
@@ -0,0 +1,260 @@
+            raise CircuitError("All modes must be measured.")
+
+        # Check circuit begins with two mode squeezers
+        # --------------------------------------------
+        A, B, C = group_operations(seq, lambda x: isinstance(x, ops.S2gate))
+
+        print([o.op for o in A])
+
+        if A:
+            raise CircuitError("Circuits must start with four S2gates.")
+
+        # get circuit registers
+        regrefs = {q for cmd in B for q in cmd.reg}
+
+        if len(regrefs) != self.modes:
+            raise CircuitError("S2gates do not appear on the correct modes.")
+
+        # Compile the unitary: combine and then decompose all unitaries
+        # -------------------------------------------------------------
+        A, B, C = group_operations(seq, lambda x: isinstance(x, (ops.Rgate, ops.BSgate)))
+
+        # begin unitary lists for mode [0, 1, 2, 3] and modes [4, 5, 6, 7] with
+        # two identity matrices. This is because multi_dot requires
+        # at least two matrices in the list.
+        U_list0 = [np.identity(self.modes // 2, dtype=np.complex128)] * 2
+        U_list4 = [np.identity(self.modes // 2, dtype=np.complex128)] * 2
+
+        if not B:
+            # no interferometer was applied
+            A, B, C = group_operations(seq, lambda x: isinstance(x, ops.S2gate))
+            A = B  # move the S2gates to A
+        else:
+            for cmd in B:
+                # calculate the unitary matrix representing each
+                # rotation gate and each beamsplitter
+                # Note: this is done separately on modes [0, 1, 2, 3]
+                # and modes [4, 5, 6, 7]
+                modes = [i.ind for i in cmd.reg]
+                params = par_evaluate(cmd.op.p)
+                U = np.identity(self.modes // 2, dtype=np.complex128)
+
+                if isinstance(cmd.op, ops.Rgate):
+                    m = modes[0]
+                    U[m % 4, m % 4] = np.exp(1j * params[0])
+
+                elif isinstance(cmd.op, ops.BSgate):
+                    m, n = modes
+
+                    t = np.cos(params[0])
+                    r = np.exp(1j * params[1]) * np.sin(params[0])
+
+                    U[m % 4, m % 4] = t
+                    U[m % 4, n % 4] = -np.conj(r)
+                    U[n % 4, m % 4] = r
+                    U[n % 4, n % 4] = t
+
+                if set(modes).issubset({0, 1, 2, 3}):
+                    U_list0.insert(0, U)
+                elif set(modes).issubset({4, 5, 6, 7}):
+                    U_list4.insert(0, U)
+                else:
+                    raise CircuitError(
+                        "Unitary must be applied separately to modes [0, 1, 2, 3] and modes [4, 5, 6, 7]."
+                    )
+
+        # multiply all unitaries together, to get the final
+        # unitary representation on modes [0, 1] and [2, 3].
+        U0 = multi_dot(U_list0)
+        U4 = multi_dot(U_list4)
+
+        # check unitaries are equal
```

Comment:
The reason a separate U0 and U4 is created is so that we can call two separate `ops.Interferometer` calls. Currently, the MZ decomposition on a single `U` matrix will result in (identity) beamsplitters and rotation gates occurring across the modes that shouldn't be connected. Kind of annoying.

Reply:
Not sure I follow what the issue is here, but in any case, can't we just create one matrix `U` and then call `ops.Interferometer` twice with the same argument? (`U0` and `U4` are effectively the same argument already, but taking up twice as much memory)

Reply:
Oh I see what you mean. I thought you meant one `Interferometer` call.

Reply:
I had a think about this, and for now (because of the way SF is written), it is possible for a user to apply separate unitaries to modes [0,1,2,3] and [4,5,6,7]. For example:

```python
with prog.context as q:
    ops.S2gate(0.2, 0) | (q[0], q[4])
    ops.S2gate(0.3, 0) | (q[1], q[5])
    ops.S2gate(0.1, 0) | (q[2], q[6])
    ops.S2gate(0.4, 0) | (q[3], q[7])

    # unitary on modes [0,1,2,3]
    ops.Interferometer(U) | (q[0], q[1], q[2], q[3])

    # unitary on modes [4,5,6,7]
    ops.Interferometer(U) | (q[4], q[5], q[6], q[7])
    ops.BSgate() | (q[2], q[3])

    # measurements
    ops.MeasureFock() | q
```
Since the (tunable) squeezers couple these two sets of modes, we need to create this as a 8 mode program (not a 4 mode program), so I think this is unavoidable for now.

Reply:
But this program doesn't fit the circuitspec anyway, so why are we considering it?###

Reply:
nevermind, I think I see it now